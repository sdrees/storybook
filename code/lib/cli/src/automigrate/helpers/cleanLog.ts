import { EOL } from 'os';

export const cleanLog = (str: string) =>
  str
    // remove chalk ANSI colors
    // eslint-disable-next-line no-control-regex
    .replace(/\u001b\[.*?m/g, '')
    // fix boxen output
    .replace(/╮│/g, '╮\n│')
    .replace(/││/g, '│\n│')
    .replace(/│╰/g, '│\n╰')
    .replace(/⚠️ {2}failed to check/g, `${EOL}⚠️  failed to check`);
