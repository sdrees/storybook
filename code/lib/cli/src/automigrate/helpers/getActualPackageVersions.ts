import * as fs from 'fs-extra';
import path from 'path';

export const getActualPackageVersions = async (packages: string[]) => {
  return Promise.all(packages.map(getActualPackageVersion));
};

export const getActualPackageVersion = async (packageName: string) => {
  try {
    const packageJson = await getActualPackageJson(packageName);
    return {
      name: packageName,
      version: packageJson.version,
    };
  } catch (err) {
    return { name: packageName, version: null };
  }
};

export const getActualPackageJson = async (packageName: string) => {
  const resolvedPackageJson = require.resolve(path.join(packageName, 'package.json'), {
    paths: [process.cwd()],
  });
  const packageJson = await fs.readJson(resolvedPackageJson);
  return packageJson;
};
