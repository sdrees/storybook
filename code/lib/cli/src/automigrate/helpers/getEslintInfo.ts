import fse from 'fs-extra';

export const SUPPORTED_ESLINT_EXTENSIONS = ['js', 'cjs'];
const UNSUPPORTED_ESLINT_EXTENSIONS = ['yaml', 'yml', 'json'];

export const findEslintFile = () => {
  const filePrefix = '.eslintrc';
  const unsupportedExtension = UNSUPPORTED_ESLINT_EXTENSIONS.find((ext: string) =>
    fse.existsSync(`${filePrefix}.${ext}`)
  );

  if (unsupportedExtension) {
    throw new Error(unsupportedExtension);
  }

  const extension = SUPPORTED_ESLINT_EXTENSIONS.find((ext: string) =>
    fse.existsSync(`${filePrefix}.${ext}`)
  );
  return extension ? `${filePrefix}.${extension}` : null;
};
