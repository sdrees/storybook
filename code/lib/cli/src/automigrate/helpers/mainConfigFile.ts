import {
  getStorybookInfo,
  loadMainConfig,
  rendererPackages,
  frameworkPackages,
  builderPackages,
} from '@storybook/core-common';
import type { StorybookConfigRaw, StorybookConfig } from '@storybook/types';
import type { ConfigFile } from '@storybook/csf-tools';
import { readConfig, writeConfig as writeConfigFile } from '@storybook/csf-tools';
import chalk from 'chalk';
import dedent from 'ts-dedent';
import path from 'path';
import type { JsPackageManager } from '../../js-package-manager';
import { getStorybookVersion } from '../../utils';

const logger = console;

/**
 * Given a Storybook configuration object, retrieves the package name or file path of the framework.
 * @param mainConfig - The main Storybook configuration object to lookup.
 * @returns - The package name of the framework. If not found, returns null.
 */
export const getFrameworkPackageName = (mainConfig?: StorybookConfigRaw) => {
  const packageNameOrPath =
    typeof mainConfig?.framework === 'string' ? mainConfig.framework : mainConfig?.framework?.name;

  if (!packageNameOrPath) {
    return null;
  }

  const normalizedPath = path.normalize(packageNameOrPath).replace(new RegExp(/\\/, 'g'), '/');

  return (
    Object.keys(frameworkPackages).find((pkg) => normalizedPath.endsWith(pkg)) || packageNameOrPath
  );
};

/**
 * Given a Storybook configuration object, retrieves the package name or file path of the builder.
 * @param mainConfig - The main Storybook configuration object to lookup.
 * @returns - The package name of the builder. If not found, returns null.
 */
export const getBuilderPackageName = (mainConfig?: StorybookConfigRaw) => {
  const packageNameOrPath =
    typeof mainConfig?.core?.builder === 'string'
      ? mainConfig.core.builder
      : mainConfig?.core?.builder?.name;

  if (!packageNameOrPath) {
    return null;
  }

  const normalizedPath = path.normalize(packageNameOrPath).replace(new RegExp(/\\/, 'g'), '/');

  return builderPackages.find((pkg) => normalizedPath.endsWith(pkg)) || packageNameOrPath;
};

/**
 * Returns a renderer package name given a framework package name.
 * @param frameworkPackageName - The package name of the framework to lookup.
 * @returns - The corresponding package name in `rendererPackages`. If not found, returns null.
 */
export const getRendererPackageNameFromFramework = (frameworkPackageName: string) => {
  if (frameworkPackageName) {
    if (Object.keys(rendererPackages).includes(frameworkPackageName)) {
      // at some point in 6.4 we introduced a framework field, but filled with a renderer package
      return frameworkPackageName;
    }

    if (Object.values(rendererPackages).includes(frameworkPackageName)) {
      // for scenarios where the value is e.g. "react" instead of "@storybook/react"
      return Object.keys(rendererPackages).find(
        (k) => rendererPackages[k] === frameworkPackageName
      );
    }
  }

  return null;
};

export const getStorybookData = async ({
  packageManager,
  configDir: userDefinedConfigDir,
}: {
  packageManager: JsPackageManager;
  configDir: string;
}) => {
  const packageJson = await packageManager.retrievePackageJson();
  const {
    mainConfig: mainConfigPath,
    version: storybookVersionSpecifier,
    configDir: configDirFromScript,
    previewConfig: previewConfigPath,
  } = getStorybookInfo(packageJson, userDefinedConfigDir);
  const storybookVersion = await getStorybookVersion(packageManager);

  const configDir = userDefinedConfigDir || configDirFromScript || '.storybook';

  let mainConfig: StorybookConfigRaw;
  try {
    mainConfig = (await loadMainConfig({ configDir, noCache: true })) as StorybookConfigRaw;
  } catch (err) {
    throw new Error(
      dedent`Unable to find or evaluate ${chalk.blue(mainConfigPath)}: ${err.message}`
    );
  }

  return {
    configDir,
    mainConfig,
    storybookVersionSpecifier,
    storybookVersion,
    mainConfigPath,
    previewConfigPath,
  };
};
export type GetStorybookData = typeof getStorybookData;

/**
 * A helper function to safely read and write the main config file. At the end of the callback, main.js will be overwritten.
 * If it fails, it will handle the error and log a message to the user explaining what to do.
 *
 * It receives a mainConfigPath and a callback
 * which will have access to utilities to manipulate main.js.
 *
 * @example
 * ```ts
 * await safeWriteMain({ mainConfigPath, dryRun }, async ({ main }) => {
 *  // manipulate main.js here
 * });
 * ```
 */
export const updateMainConfig = async (
  { mainConfigPath, dryRun }: { mainConfigPath: string; dryRun: boolean },
  callback: (main: ConfigFile) => Promise<void> | void
) => {
  try {
    const main = await readConfig(mainConfigPath);
    await callback(main);
    if (!dryRun) {
      await writeConfigFile(main);
    }
  } catch (e) {
    logger.info(
      `❌ The migration failed to update your ${chalk.blue(
        mainConfigPath
      )} on your behalf because of the following error:
        ${e}\n`
    );
    logger.info(
      `⚠️ Storybook automigrations are based on AST parsing and it's possible that your ${chalk.blue(
        mainConfigPath
      )} file contains a non-standard format (e.g. your export is not an object) or that there was an error when parsing dynamic values (e.g. "require" calls, or usage of environment variables). When your main config is non-standard, automigrations are unfortunately not possible. Please follow the instructions given previously and follow the documentation to make the updates manually.`
    );
  }
};

export const getAddonNames = (mainConfig: StorybookConfig): string[] => {
  const addons = mainConfig.addons || [];
  const addonList = addons.map((addon) => {
    let name = '';
    if (typeof addon === 'string') {
      name = addon;
    } else if (typeof addon === 'object') {
      name = addon.name;
    }

    if (name.startsWith('.')) {
      return undefined;
    }

    return name
      .replace(/\/dist\/.*/, '')
      .replace(/\.[mc]?[tj]?s[x]?$/, '')
      .replace(/\/register$/, '')
      .replace(/\/manager$/, '')
      .replace(/\/preset$/, '');
  });

  return addonList.filter(Boolean);
};
