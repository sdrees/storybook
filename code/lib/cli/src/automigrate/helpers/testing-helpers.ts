import type { JsPackageManager, PackageJson } from '../../js-package-manager';

jest.mock('./mainConfigFile', () => ({
  ...jest.requireActual('./mainConfigFile'),
  getStorybookData: jest.fn(),
}));

jest.mock('@storybook/core-common', () => ({
  ...jest.requireActual('@storybook/core-common'),
  loadMainConfig: jest.fn(),
}));

export const makePackageManager = (packageJson: PackageJson) => {
  const { dependencies = {}, devDependencies = {}, peerDependencies = {} } = packageJson;
  return {
    retrievePackageJson: async () => ({ dependencies: {}, devDependencies: {}, ...packageJson }),
    getAllDependencies: async () => ({
      ...dependencies,
      ...devDependencies,
      ...peerDependencies,
    }),
  } as JsPackageManager;
};
