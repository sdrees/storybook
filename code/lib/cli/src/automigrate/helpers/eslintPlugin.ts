import fse, { readFile, readJson, writeJson } from 'fs-extra';

import { dedent } from 'ts-dedent';
import detectIndent from 'detect-indent';
import { readConfig, writeConfig } from '@storybook/csf-tools';
import prompts from 'prompts';
import chalk from 'chalk';
import type { JsPackageManager } from '../../js-package-manager';
import { paddedLog } from '../../helpers';

export const SUPPORTED_ESLINT_EXTENSIONS = ['js', 'cjs', 'json'];
const UNSUPPORTED_ESLINT_EXTENSIONS = ['yaml', 'yml'];

export const findEslintFile = () => {
  const filePrefix = '.eslintrc';
  const unsupportedExtension = UNSUPPORTED_ESLINT_EXTENSIONS.find((ext: string) =>
    fse.existsSync(`${filePrefix}.${ext}`)
  );

  if (unsupportedExtension) {
    throw new Error(unsupportedExtension);
  }

  const extension = SUPPORTED_ESLINT_EXTENSIONS.find((ext: string) =>
    fse.existsSync(`${filePrefix}.${ext}`)
  );
  return extension ? `${filePrefix}.${extension}` : null;
};

export async function extractEslintInfo(packageManager: JsPackageManager): Promise<{
  hasEslint: boolean;
  isStorybookPluginInstalled: boolean;
  eslintConfigFile: string | null;
}> {
  const allDependencies = await packageManager.getAllDependencies();
  const packageJson = await packageManager.retrievePackageJson();
  let eslintConfigFile: string | null = null;

  try {
    eslintConfigFile = findEslintFile();
  } catch (err) {
    //
  }

  const isStorybookPluginInstalled = !!allDependencies['eslint-plugin-storybook'];
  const hasEslint = allDependencies.eslint || eslintConfigFile || packageJson.eslintConfig;
  return { hasEslint, isStorybookPluginInstalled, eslintConfigFile };
}

export async function configureEslintPlugin(eslintFile: string, packageManager: JsPackageManager) {
  if (eslintFile) {
    paddedLog(`Configuring Storybook ESLint plugin at ${eslintFile}`);
    if (eslintFile.endsWith('json')) {
      const eslintConfig = (await readJson(eslintFile)) as { extends?: string[] };
      const existingConfigValue = Array.isArray(eslintConfig.extends)
        ? eslintConfig.extends
        : [eslintConfig.extends].filter(Boolean);
      eslintConfig.extends = [...(existingConfigValue || []), 'plugin:storybook/recommended'];

      const eslintFileContents = await readFile(eslintFile, 'utf8');
      const spaces = detectIndent(eslintFileContents).amount || 2;
      await writeJson(eslintFile, eslintConfig, { spaces });
    } else {
      const eslint = await readConfig(eslintFile);
      const extendsConfig = eslint.getFieldValue(['extends']) || [];
      const existingConfigValue = Array.isArray(extendsConfig)
        ? extendsConfig
        : [extendsConfig].filter(Boolean);
      eslint.setFieldValue(
        ['extends'],
        [...(existingConfigValue || []), 'plugin:storybook/recommended']
      );

      await writeConfig(eslint);
    }
  } else {
    paddedLog(`Configuring eslint-plugin-storybook in your package.json`);
    const packageJson = await packageManager.retrievePackageJson();
    await packageManager.writePackageJson({
      ...packageJson,
      eslintConfig: {
        ...packageJson.eslintConfig,
        extends: [...(packageJson.eslintConfig?.extends || []), 'plugin:storybook/recommended'],
      },
    });
  }
}

export const suggestESLintPlugin = async (): Promise<boolean> => {
  const { shouldInstall } = await prompts({
    type: 'confirm',
    name: 'shouldInstall',
    message: dedent`
        We have detected that you're using ESLint. Storybook provides a plugin that gives the best experience with Storybook and helps follow best practices: ${chalk.yellow(
          'https://github.com/storybookjs/eslint-plugin-storybook#readme'
        )}

        Would you like to install it?
      `,
    initial: true,
  });

  return shouldInstall;
};
