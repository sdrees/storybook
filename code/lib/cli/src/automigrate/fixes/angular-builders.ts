import { dedent } from 'ts-dedent';
import semver from 'semver';
import type { StorybookConfig } from '@storybook/types';
import chalk from 'chalk';
import prompts from 'prompts';
import type { Fix } from '../types';
import { isNxProject } from '../../helpers';
import { AngularJSON } from '../../generators/ANGULAR/helpers';
import type { JsPackageManager } from '../../js-package-manager';
import { getFrameworkPackageName } from '../helpers/mainConfigFile';

interface AngularBuildersRunOptions {
  mainConfig: StorybookConfig;
  packageManager: JsPackageManager;
}

export const angularBuilders: Fix<AngularBuildersRunOptions> = {
  id: 'angular-builders',

  async check({ packageManager, mainConfig }) {
    const angularVersion = await packageManager.getPackageVersion('@angular/core');

    const framewworkPackageName = getFrameworkPackageName(mainConfig);

    // Skip in case of NX
    if (
      !angularVersion ||
      (await isNxProject()) ||
      framewworkPackageName !== '@storybook/angular'
    ) {
      return null;
    }

    if (semver.lt(angularVersion, '14.0.0')) {
      throw new Error(dedent`
      ‚ùå Your project uses Angular < 14.0.0. Storybook 7.0 for Angular requires Angular 14.0.0 or higher. 
      Please upgrade your Angular version to at least version 14.0.0 to use Storybook 7.0 in your project.
      `);
    }

    const angularJSON = new AngularJSON();

    const { hasStorybookBuilder } = angularJSON;

    // skip if workspace has already one or more Storybook builder
    if (hasStorybookBuilder) {
      return null;
    }

    if (!angularJSON.rootProject && Object.keys(angularJSON.projects).length > 1) {
      return null;
    }

    return {
      mainConfig,
      packageManager,
    };
  },

  prompt() {
    return dedent`
      We have detected that your project does not use the Storybook Angular builder yet. In Storybook 6.4 we deprecated calling Storybook directly (npm run storybook) for Angular. In Storybook 7.0, we've removed it entirely.
      
      In order to use the Storybook Angular builder, we need to add a few entries to your angular.json file. Additionally, we will add the @compodoc/compodoc package to your devDependencies if you want and we will add a few scripts to your package.json file.

      Also feel free to remove the Compodoc script from your package.json file if you don't use it apart from Storybook anymore. Storybook uses Compodoc internally and you don't have to call in separately anymore.

      Read more about the Angular builder here: ${chalk.yellow(
        'https://github.com/storybookjs/storybook/tree/next/code/frameworks/angular#how-do-i-migrate-to-an-angular-storybook-builder'
      )}
    `;
  },

  async run({ result }) {
    const angularJSON = new AngularJSON();
    const { packageManager } = result;

    const { useCompoDoc } = await prompts({
      type: 'confirm',
      name: 'useCompoDoc',
      message: 'Have you set up compodoc in Storybook previously?',
    });

    const angularProjectName = await angularJSON.getProjectName();

    angularJSON.addStorybookEntries({
      angularProjectName,
      storybookFolder: '.storybook',
      useCompodoc: useCompoDoc,
      root: '.',
    });

    angularJSON.write();

    await packageManager.addScripts({
      storybook: `ng run ${angularProjectName}:storybook`,
      'build-storybook': `ng run ${angularProjectName}:build-storybook`,
    });
  },
};
