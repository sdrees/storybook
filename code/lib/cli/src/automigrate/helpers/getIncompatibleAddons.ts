import type { StorybookConfig } from '@storybook/types';
import semver from 'semver';
import { getActualPackageVersions } from './getActualPackageVersions';
import { getAddonNames } from './mainConfigFile';

export const getIncompatibleAddons = async (mainConfig: StorybookConfig) => {
  // TODO: Keep this up to date with https://github.com/storybookjs/storybook/issues/20529 in case more addons get added
  const incompatibleList = {
    '@storybook/addon-knobs': '6.4.0',
    '@storybook/addon-postcss': '2.0.0',
    'storybook-addon-next-router': '4.0.2',
    'storybook-addon-outline': '1.4.2', // (deprecated)
    '@storybook/addon-info': '5.3.21',
    'storybook-addon-designs': '6.3.1',
    'storybook-addon-next': '1.7.0', // (deprecated)'
    'storybook-docs-toc': '1.7.0',
    '@storybook/addon-google-analytics': '6.2.9',
    'storybook-addon-pseudo-states': '1.15.5',
    'storybook-dark-mode': '2.1.1',
    'storybook-addon-gatsby': '0.0.5',
    '@etchteam/storybook-addon-css-variables-theme': '1.4.0',
    '@storybook/addon-cssresources': '6.2.9',
    'storybook-addon-grid': '0.3.1',
    'storybook-multilevel-sort': '1.2.0',
    'storybook-addon-i18next': '1.3.0',
    'storybook-source-link': '2.0.8',
    'babel-plugin-storybook-csf-title': '2.1.0',
    '@urql/storybook-addon': '2.0.1',
    'storybook-addon-intl': '2.4.1',
    'storybook-addon-mock': '3.2.0',
    '@chakra-ui/storybook-addon': '4.0.16',
    'storybook-mobile-addon': '1.0.2',
    '@storybook/addon-queryparams': '6.2.9',
  };

  const addons = getAddonNames(mainConfig).filter((addon) => addon in incompatibleList);

  if (addons.length === 0) {
    return [];
  }

  const addonVersions = await getActualPackageVersions(addons);

  const incompatibleAddons: { name: string; version: string }[] = [];
  addonVersions.forEach(
    ({
      name,
      version: installedVersion,
    }: {
      name: keyof typeof incompatibleList;
      version: string;
    }) => {
      if (installedVersion === null) return;

      const addonVersion = incompatibleList[name];
      try {
        if (semver.lte(semver.coerce(installedVersion), semver.coerce(addonVersion))) {
          incompatibleAddons.push({ name, version: installedVersion });
        }
      } catch (err) {
        // we tried our best but if we can't compare, we just no-op for that addon
      }
    }
  );

  return incompatibleAddons;
};
