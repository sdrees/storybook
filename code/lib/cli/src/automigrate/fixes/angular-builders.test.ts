import type { StorybookConfig } from '@storybook/types';
import { angularBuilders } from './angular-builders';
import * as helpers from '../../helpers';
import * as angularHelpers from '../../generators/ANGULAR/helpers';
import type { JsPackageManager } from '../../js-package-manager';

const checkAngularBuilders = async ({
  packageManager,
  mainConfig = {},
  storybookVersion = '7.0.0',
}: {
  packageManager: Partial<JsPackageManager>;
  mainConfig?: Partial<StorybookConfig>;
  storybookVersion?: string;
}) => {
  return angularBuilders.check({
    packageManager: packageManager as any,
    storybookVersion,
    mainConfig: mainConfig as any,
  });
};

jest.mock('../../helpers', () => ({
  ...jest.requireActual('../../helpers'),
  isNxProject: jest.fn(),
}));

jest.mock('../../generators/ANGULAR/helpers', () => ({
  ...jest.requireActual('../../generators/ANGULAR/helpers'),
  AngularJSON: jest.fn(),
}));

describe('is Nx project', () => {
  beforeEach(() => {
    (helpers.isNxProject as any as jest.SpyInstance).mockResolvedValue(true);
  });

  const packageManager = {
    getPackageVersion: jest.fn().mockImplementation((packageName) => {
      if (packageName === '@angular/core') {
        return '12.0.0';
      }

      return null;
    }),
  } as Partial<JsPackageManager>;

  it('should return null', async () => {
    await expect(checkAngularBuilders({ packageManager })).resolves.toBeNull();
  });
});

describe('is not Nx project', () => {
  beforeEach(() => {
    (helpers.isNxProject as any as jest.SpyInstance).mockResolvedValue(false);
  });

  describe('angular builders', () => {
    afterEach(jest.restoreAllMocks);

    describe('Angular not found', () => {
      const packageManager = {
        getPackageVersion: jest.fn().mockReturnValue(null),
      } as Partial<JsPackageManager>;

      it('should return null', async () => {
        await expect(checkAngularBuilders({ packageManager })).resolves.toBeNull();
      });
    });

    describe('Angular < 14.0.0', () => {
      const packageManager = {
        getPackageVersion: (packageName: string) => {
          if (packageName === '@angular/core') {
            return Promise.resolve('12.0.0');
          }

          return null;
        },
      } as Partial<JsPackageManager>;

      it('should throw an Error', async () => {
        await expect(
          checkAngularBuilders({ packageManager, mainConfig: { framework: '@storybook/angular' } })
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });

    describe('Angular >= 14.0.0', () => {
      const packageManager = {
        getPackageVersion: (packageName) => {
          if (packageName === '@angular/core') {
            return Promise.resolve('15.0.0');
          }

          return null;
        },
      } as Partial<JsPackageManager>;

      describe('has one Storybook builder defined', () => {
        beforeEach(() => {
          // Mock AngularJSON.constructor
          (angularHelpers.AngularJSON as jest.Mock).mockImplementation(() => ({
            hasStorybookBuilder: true,
          }));
        });

        it('should return null', async () => {
          await expect(
            checkAngularBuilders({
              packageManager,
              mainConfig: { framework: '@storybook/angular' },
            })
          ).resolves.toBeNull();
        });
      });

      describe('has multiple projects without root project defined', () => {
        beforeEach(() => {
          // Mock AngularJSON.constructor
          (angularHelpers.AngularJSON as jest.Mock).mockImplementation(() => ({
            hasStorybookBuilder: false,
            projects: {
              project1: { root: 'project1', architect: {} },
              project2: { root: 'project2', architect: {} },
            },
            rootProject: null,
          }));
        });

        it('should return null', async () => {
          await expect(
            checkAngularBuilders({
              packageManager,
              mainConfig: { framework: '@storybook/angular' },
            })
          ).resolves.toBeNull();
        });
      });

      describe('has one project', () => {
        beforeEach(() => {
          // Mock AngularJSON.constructor
          (angularHelpers.AngularJSON as jest.Mock).mockImplementation(() => ({
            hasStorybookBuilder: false,
            projects: {
              project1: { root: 'project1', architect: {} },
            },
            rootProject: 'project1',
          }));
        });

        it('should proceed and return data', async () => {
          await expect(
            checkAngularBuilders({
              packageManager,
              mainConfig: { framework: '@storybook/angular' },
            })
          ).resolves.toMatchObject({
            mainConfig: {},
            packageManager: {},
          });
        });
      });
    });
  });
});
