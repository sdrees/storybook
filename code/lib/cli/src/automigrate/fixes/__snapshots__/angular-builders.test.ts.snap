// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is not Nx project angular builders Angular < 15.0.0 should throw an Error 1`] = `
"❌ Your project uses Angular < 15.0.0. Storybook 8.0 for Angular requires Angular 15.0.0 or higher. 
Please upgrade your Angular version to at least version 15.0.0 to use Storybook 8.0 in your project."
`;
